stablehlo/reference/Element.h

```diff
@@ -52,7 +62,7 @@ class Element {
 
  private:
   Type type_;
-  Attribute value_;
+  std::variant<APInt, APFloat, std::pair<APFloat, APFloat>> value_;
```

Comment:
Why can't this be `std::variant<APInt, APFloat, std::complex<APFloat>>`? This works for me.

```
diff --git a/stablehlo/reference/Element.h b/stablehlo/reference/Element.h
index dfc454eca2..61f07ab01c 100644
--- a/stablehlo/reference/Element.h
+++ b/stablehlo/reference/Element.h
@@ -32,8 +32,7 @@ class Element {
   /// @{
   Element(Type type, APInt val) : type_(type), value_(val) {}
   Element(Type type, APFloat val) : type_(type), value_(val) {}
-  Element(Type type, std::complex<APFloat> val)
-      : type_(type), value_(std::make_pair(val.real(), val.imag())) {}
+  Element(Type type, std::complex<APFloat> val) : type_(type), value_(val) {}

   Element(const Element &other) = default;
   /// @}
@@ -47,8 +46,7 @@ class Element {
   APFloat getFloatValue() const { return std::get<APFloat>(value_); }
   APInt getIntegerValue() const { return std::get<APInt>(value_); }
   std::complex<APFloat> getComplexValue() const {
-    auto floatPair = std::get<std::pair<APFloat, APFloat>>(value_);
-    return std::complex<APFloat>(floatPair.first, floatPair.second);
+    return std::get<std::complex<APFloat>>(value_);
   }

   /// Overloaded + operator.
@@ -62,7 +60,7 @@ class Element {

  private:
   Type type_;
-  std::variant<APInt, APFloat, std::pair<APFloat, APFloat>> value_;
+  std::variant<APInt, APFloat, std::complex<APFloat>> value_;
 };
```

Reply:
Strange that I am still getting the same error with the above change.

```
In file included from /usr/local/google/home/sdasgup/Github/stablehlo/stablehlo/reference/Ops.cpp:16:                                                                                                                           
In file included from /usr/local/google/home/sdasgup/Github/stablehlo/stablehlo/reference/Ops.h:19:                                                                                                                             
In file included from /usr/local/google/home/sdasgup/Github/stablehlo/llvm-project/mlir/include/mlir/IR/BuiltinAttributes.h:12:                                                                                                 
In file included from /usr/local/google/home/sdasgup/Github/stablehlo/llvm-project/mlir/include/mlir/IR/BuiltinAttributeInterfaces.h:19:                                                                                        
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/complex:135:75: error: calling a private constructor of class 'llvm::APFloat'                                                                                
      _GLIBCXX_CONSTEXPR complex(const _Tp& __r = _Tp(), const _Tp& __i = _Tp())                                                                                                                                                
                                                                          ^                                                                                                                                                     
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:774:36: note: in instantiation of default function argument expression for 'complex<llvm::APFloat>' required here                                    
                      void_t<decltype(_Arr<_Ti>{{std::declval<_Tp>()}})>>                                                                                                                                                       
                                                 ^                                                                                                                                                                              
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:786:7: note: during template argument deduction for class template partial specialization '_Build_FUN<_Ind, _Tp, _Ti>' [with _Ind = 2, _Tp = llvm::AP
Float &, _Ti = std::complex<llvm::APFloat>]                                                                                                                                                                                     
    : _Build_FUN<_Ind, _Tp, _Ti>...                                                                                                                                                                                             
      ^                                                                                                                                                                                                                         
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:786:7: note: in instantiation of template class 'std::__detail::__variant::_Build_FUN<2, llvm::APFloat &, std::complex<llvm::APFloat>>' requested her
e                                                                                                                                                                                                                               
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:795:18: note: in instantiation of template class 'std::__detail::__variant::_Build_FUNs<llvm::APFloat &, std::variant<llvm::APInt, llvm::APFloat, std
::complex<llvm::APFloat>>, std::integer_sequence<unsigned long, 0, 1, 2>>' requested here                                                                                                                                       
      = decltype(_Build_FUNs<_Tp, _Variant>::_S_fun(std::declval<_Tp>()));                                                                                                                                                      
                 ^                                                                                                                                                                                                              
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:804:51: note: in instantiation of template type alias '_FUN_type' requested here                                                                     
    struct __accepted_index<_Tp, _Variant, void_t<_FUN_type<_Tp, _Variant>>>                                                                                                                                                    
                                                  ^                                                                                                                                                                             
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:1376:27: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)                                                        
          = __detail::__variant::__accepted_index<_Tp, variant>::value;                                                                                                                                                         
                                 ^                                                                                                                                                                                              
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:1376:27: note: in instantiation of template class 'std::__detail::__variant::__accepted_index<llvm::APFloat &, std::variant<llvm::APInt, llvm::APFloa
t, std::complex<llvm::APFloat>>>' requested here                                                                                                                                                                                
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:1382:36: note: in instantiation of static data member 'std::variant<llvm::APInt, llvm::APFloat, std::complex<llvm::APFloat>>::__accepted_index<llvm::
APFloat &>' requested here                                                                                                                                                                                                      
        using __accepted_type = __to_type<__accepted_index<_Tp>>;                                                                                                                                                               
                                          ^                                                                                                                                                                                     
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:1412:24: note: in instantiation of template type alias '__accepted_type' requested here                                                              
               typename _Tj = __accepted_type<_Tp&&>,                                                                                                                                                                           
                              ^                                                                                                                                                                                                 
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/variant:1416:2: note: in instantiation of default argument for 'variant<llvm::APFloat &, void, void>' required here                                          
        variant(_Tp&& __t)                                                                                                                                                                                                      
        ^~~~~~~~~~~~~~~~~~                                                                                                                                                                                                      
/usr/local/google/home/sdasgup/Github/stablehlo/stablehlo/reference/Element.h:35:59: note: while substituting deduced template arguments into function template 'variant' [with _Tp = llvm::APFloat &, $1 = (no value), $2 = (no
 value), _Tj = (no value), $4 = (no value)]                                                                                                                                                                                     
  Element(Type type, APFloat value) : type_(type), value_(value) {}                                                                                                                                                             
                                                          ^                                                                                                                                                                     
/usr/local/google/home/sdasgup/Github/stablehlo/llvm-project/llvm/include/llvm/ADT/APFloat.h:844:3: note: implicitly declared private here                                                                                      
  APFloat() : U(IEEEdouble()) {                                                                                                                                                                                                 
  ^                             
```

Reply:
Interesting! What compiler are you using? Here's a pull request that works on my machine: https://github.com/sdasgup3/stablehlo/pull/2 - just so that we're sure we're looking at the same code.

Reply:
```
Debian clang version 13.0.1-7
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

Unfortunately the error persists.